### PR TITLE
Globalization - add analyzers

### PR DESCRIPTION
**Pull request by LucGenetier**

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/Delegation/DelegationStrategies/OpDelegationStrategy.cs [hunk 1]
```diff
@@ -100,8 +101,12 @@ private bool IsSupportedNode(TexlNode node, OperationCapabilityMetadata metadata
                 // Set(Names, ["Foo", Bar"]); Filter(Accounts, 'Account Name' in Names) - Using variable of type table
                 // ClearCollect(Names, Accounts); Filter(Accounts, 'Account Name' in Names.'Account Name') - using column from collection.
                 // This won't be delegated - Filter(Accounts, 'Account Name' in Accounts.'Account Name') as Accounts.'Account Name' is async.
-                if (isRHSNode && binding.Document.Properties.EnabledFeatures.IsEnhancedDelegationEnabled
-                    && (opDelStrategy as BinaryOpDelegationStrategy)?.Op == BinaryOp.In && !binding.IsAsync(node) && binding.GetType(node).IsTable && binding.GetType(node).IsColumn &&
+                if (isRHSNode &&
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/Delegation/DelegationStrategies/OpDelegationStrategy.cs:104]: why this change?

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 2]
```diff
@@ -199,8 +199,11 @@ static TexlLexer()
         public static TexlLexer GetLocalizedInstance(CultureInfo culture)
         {
             culture ??= CultureInfo.InvariantCulture;
+
+            // Number decimal separator can be a dot (.), comma (,), arabic comma (Unicode 0x66B) 
+            // .Net core 3.1 doesn't support arabic comma and defines '/' as decimal separator for Persian/Farsi (fa, fa-IR) which is probably a bug
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:204]: >// [](http://example.com/codeflow?start=12&length=2)

Interesting ... Is there something to followup here?
Can we craft a bug

File: src/libraries/Microsoft.PowerFx.Core/Parser/ParseResult.cs [hunk 3]
```diff
@@ -106,6 +100,8 @@ internal ParseResult(TexlNode root, List<TexlError> errors, bool hasError, List<
             After = after;
 
             Text = text;
+
+            ErrorMessageLocale = errorMessageLocale ?? CultureInfo.CurrentCulture;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Parser/ParseResult.cs:104]: > ?? CultureInfo.CurrentCulture [](http://example.com/codeflow?start=51&length=30)

why?
Changing from null to non-null could be significant.

File: src/libraries/Microsoft.PowerFx.Core/Public/ExpressionError.cs [hunk 4]
```diff
@@ -147,7 +147,7 @@ internal static IEnumerable<ExpressionError> New(IEnumerable<IDocumentError> err
             }
             else
             {
-                return errors.Select(x => ExpressionError.New(x)).ToArray();
+                return errors.Select(x => ExpressionError.New(x, CultureInfo.CurrentCulture)).ToArray();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/ExpressionError.cs:150]: >CultureInfo.CurrentCulture [](http://example.com/codeflow?start=65&length=26)

Why this change? We want to avoid  using CultureInfo.CurrentCulture.  

We want localization to be lazy (at point of reading the message), not point of creating the message.

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/NumberValue.cs [hunk 5]
```diff
@@ -22,7 +23,7 @@ public override void Visit(IValueVisitor visitor)
 
         public override void ToExpression(StringBuilder sb, FormulaValueSerializerSettings settings)
         {
-            sb.Append((Value == 0) ? "0" : Value.ToString());
+            sb.Append((Value == 0) ? "0" : Value.ToString(CultureInfo.CurrentCulture));
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Values/NumberValue.cs:26]: >CultureInfo.CurrentCulture [](http://example.com/codeflow?start=58&length=26)

No - we shouldn't use CultureInfo.CurrentCulture

Use Invariant.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs [hunk 6]
```diff
@@ -447,7 +447,7 @@ public static async ValueTask<FormulaValue> SortTable(EvalVisitor runner, EvalVi
             var arg1 = (LambdaFormulaValue)args[1];
             var arg2 = (StringValue)args[2];
 
-            var isDescending = arg2.Value.ToLower() == "descending";
+            var isDescending = arg2.Value.Equals("descending", StringComparison.OrdinalIgnoreCase);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:450]: >Equals [](http://example.com/codeflow?start=42&length=6)

these are nice improvements.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs [hunk 7]
```diff
@@ -150,7 +150,7 @@ public void New()
             var baseError = GetBaseError(TexlStrings.ErrInvalidName, "name");
             var baseErrors = new IDocumentError[] { baseError };
 
-            var errors = ExpressionError.New(baseErrors);
+            var errors = ExpressionError.New(baseErrors, CultureInfo.CurrentCulture);
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs:153]: >CultureInfo.CurrentCulture [](http://example.com/codeflow?start=57&length=26)

avoid CultureInfo.CurrentCulture

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs [hunk 8]
```diff
@@ -2,27 +2,24 @@
...
 
 namespace Microsoft.PowerFx.Interpreter.Tests
-{    
+{
     public class FileExpressionEvaluationTests : PowerFxTest
-    {        
+    {
         [InterpreterTheory]
-        [TxtFileData("ExpressionTestCases", "InterpreterExpressionTestCases", nameof(InterpreterRunner))]        
+        [TxtFileData("ExpressionTestCases", "InterpreterExpressionTestCases", nameof(InterpreterRunner), "en-US, fr-FR")]
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs:15]: >"en-US, fr-FR" [](http://example.com/codeflow?start=105&length=14)

Let's discuss.

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs (hunk 2 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:204]: Waiting on Carlos's thread he initiated with another team to know if we will soon support RTL languages like Arabic and Hebrew

File: src/libraries/Microsoft.PowerFx.Core/Parser/ParseResult.cs (hunk 3 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Parser/ParseResult.cs:104]: code simplification

File: src/libraries/Microsoft.PowerFx.Core/Public/ExpressionError.cs (hunk 4 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/ExpressionError.cs:150]: .Net analyzers detect when an API exists with a CultureInfo and if so, we need to use it

Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/ExpressionError.cs:150]: Not should we would want not to use it

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs (hunk 6 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:450]: thanks

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs (hunk 7 above)
Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs:153]: done

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs (hunk 8 above)
Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs]: for now, i'll keep "en-US" and we can expand later

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs [hunk 9]
```diff
@@ -171,7 +171,7 @@ public void Empty()
             // We don't want null IEnumerables. 
             // Null gets normalized to empty.
             var internalErrors = (IEnumerable<IDocumentError>)null;
-            var errors = ExpressionError.New(internalErrors);
+            var errors = ExpressionError.New(internalErrors, CultureInfo.CurrentCulture);
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs:174]: >CultureInfo.CurrentCulture [](http://example.com/codeflow?start=61&length=26)

"avoid CultureInfo.CurrentCulture"
Does it also apply here?

Comment (LucGenetier) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionErrorTests.cs:174]: This one is intentional otherwise there is no difference with line 178

File: src/libraries/Microsoft.PowerFx.Core/Parser/ParseResult.cs (hunk 3 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Parser/ParseResult.cs:104]: Can you elaborate?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/ExpressionTestCase.cs [hunk 10]
```diff
@@ -47,7 +48,7 @@ public static ExpressionTestCase Fail(string message)
 
         public override string ToString()
         {
-            var str = $"{Path.GetFileName(SourceFile)} : {SourceLine.ToString("000")} - {Input} = {Expected}";
+            var str = $"{Path.GetFileName(SourceFile)} : {SourceLine.ToString("000")} [{Locale}] - {Input} = {Expected}";
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/ExpressionTestCase.cs:51]: > [{Locale}]  [](http://example.com/codeflow?start=85&length=12)

This means "en-us" will show up in every test title.

**Comment by MikeStall** [comment]
>     public class TestRunnerTests : PowerFxTest

Can we ensure no changes in this file?
Any breaks here mean updates we need to make in PAClient since that also runs .txt tests.  



---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/TestRunnerTests/TestRunnerTests.cs:14 in 9d9a527. [](commit_id = 9d9a527d49a7542def8d547372a1c484e9543f4f, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TestRunner.cs [hunk 11]
```diff
@@ -193,25 +194,47 @@ public void AddFile(string thisFile)
...
+                    foreach (string locale in Locales)
                     {
-                        // Must be in different sources
-                        if (existingTest.SourceFile == test.SourceFile && existingTest.SetupHandlerName == test.SetupHandlerName)
+                        Engine e = _engines.FirstOrDefault(e => e.Config.CultureInfo.Name == locale);
+
+                        if (e == null)
                         {
-                            throw ParseError(i, $"Duplicate test cases in {Path.GetFileName(test.SourceFile)} on line {test.SourceLine} and {existingTest.SourceLine}");
+                            e = new Engine(new PowerFxConfig(new CultureInfo(locale)));
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TestRunner.cs:203]: >e = new Engine(new PowerFxConfig(new CultureInfo(locale))); [](http://example.com/codeflow?start=28&length=59)

Double checking - for other engines (like JS, SQL) - do we need to do something differently here?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TestRunner.cs [hunk 12]
```diff
@@ -193,25 +194,47 @@ public void AddFile(string thisFile)
...
-                        // Inputs are the same, but update the results.
-                        existingTest.MarkOverride(test);
-                    }
-                    else
-                    {
-                        // New test
-                        Tests.Add(test);
+                        TestCase testWithLocale = new TestCase()
+                        {
+                            Input = (locale == "en-US") ? test.Input : e.GetDisplayExpression(test.Input, null as ReadOnlySymbolTable),
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/TestRunner.cs:209]: >GetDisplayExpression [](http://example.com/codeflow?start=73&length=20)

What about?

This test case:  
`Value("123,456")`

GetDisplayExpression doesn't convert strings. 

We may have some test holes here, I added this test in a PR: 
[https://github.com/microsoft/Power-Fx/pull/1157](https://github.com/microsoft/Power-Fx/pull/1157)

**Review by MikeStall — CHANGES_REQUESTED** [review]
:clock1:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by MikeStall — APPROVED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by MikeStall** [comment]
Adding BreakingInternals  tag - this appears to have changed ordering in a way that is breaking Power Apps test.
